File: Message/src/main/java/com/yahoo/sdvornik/message/codec/ShuffleDecoder.java
package com.yahoo.sdvornik.message.codec;

import com.yahoo.sdvornik.message.DataMessageArray;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageDecoder;

import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class ShuffleDecoder extends MessageToMessageDecoder<DataMessageArray> {

    private final static ThreadLocalRandom random = ThreadLocalRandom.current();

    @Override
    protected void decode(ChannelHandlerContext ctx, DataMessageArray msg, List<Object> out) throws Exception {
        long[] a = msg.getArray();
        for(int i = a.length-1; i>=0; --i) {
            int j = random.nextInt(i+1);
            long tmp = a[j];
            a[j] = a[i];
            a[i] = tmp;
        }
        out.add(msg);
    }
}

File: shared/src/main/kotlin/com/yahoo/sdvornik/message/codec/ShuffleDecoder.kt
package com.yahoo.sdvornik.message.codec

import com.yahoo.sdvornik.message.DataMessageArray
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageDecoder
import java.util.concurrent.ThreadLocalRandom

/**
 * @author Serg Dvornik <sdvornik@yahoo.com>
 */
class ShuffleDecoder : MessageToMessageDecoder<DataMessageArray>() {

    @Throws(Exception::class)
    override fun decode(ctx: ChannelHandlerContext, msg: DataMessageArray, out: MutableList<Any>) {
        val a = msg.array
        for (i in a.indices.reversed()) {
            val j = random.nextInt(i + 1)
            val tmp = a[j]
            a[j] = a[i]
            a[i] = tmp
        }
        out.add(msg)
    }

    companion object {
        private val random = ThreadLocalRandom.current()
    }
}